File: 83 - Remove Duplicates from Sorted List.cpp
/*
 83. Remove Duplicates from Sorted List (Easy)

 Notes:
    1. fast-slow pointer - O(n)
    Use Two pointers to scan the list, if the two pointers' value are the same, move fast pointer backward, else, slow
    pointer's next points to fast pointer node, slow pointer points to fast pointer's node.

    2. One poiner - O(n)
    In the previous method, we skip the duplicate nodes, this causes memory leak.
    The better solution should be using a pointer to scan the list, if the next node value is same as the current one,
    delete the duplicate and point the current one's next to the next next one.
*/

// One pointer
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode* node = head;
        while (node && node->next)
        {
            if (node->val == node->next->val)
            {
                ListNode* temp = node->next;
                node->next = node->next->next;
                delete temp;
            }
            else
                node = node->next;
        }
        return head;
    }
};


/*
// Fast-slow pointer (Not recommended)

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head)
            return head;
        ListNode* slow = head, * fast = head;
        while (fast)
        {
            if (slow->val != fast->val)
            {
                slow->next = fast;
                slow = fast;
                fast = fast->next;
            }
            else
                fast = fast->next;
        }
        slow->next = fast;
        return head;
    }
};

*/

File: JAVA/83 - Remove Duplicates from Sorted List.java
/*
 83. Remove Duplicates from Sorted List (E)

 Approach 1 : One pointer
 Check if node's next exist, if exist, compare two nodes val and relink. 
 Time: O(n)
 Space: O(1)

 Approach 2 : Slow-fast pointer
 Compare two ptr val, if the val are different, relink the node.
 If the val are the same, skip it.
 Time: O(n)
 Space:O(1)

*/

/* Approach 1 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode node = head;
        while (node != null && node.next != null) {
            if (node.val == node.next.val)
                node.next = node.next.next;
            else node = node.next;
        }
        return head;
    }
}

/* Approach 2 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) return head;
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null) {
            if (slow.val != fast.val) {
                slow.next = fast;
                slow = slow.next;
            }
            fast = fast.next;
        }
        slow.next = null;
        return head;
    }
}